server/src/main/java/org/elasticsearch/common/time/DateFormatters.java

```diff
@@ -1585,7 +1585,7 @@ static JavaDateFormatter merge(String pattern, List<DateFormatter> formatters) {
             if (printer == null) {
                 printer = javaDateFormatter.getPrinter();
             }
-            dateTimeFormatters.add(javaDateFormatter.getParser());
+            dateTimeFormatters.addAll(javaDateFormatter.getParsers());
```

Comment:
not a fan of this merge method. It extracts relatively low level `java.time.DateTimeFormatter` when it should stick with `org.elasticsearch.common.time.DateFormatter` abstraction as long as possible. 
Possibly the roundUpBuilder should also be used inside `JavaDateFormatter constructor`?

also I suspect that roundUpBuilder will suffer from the same problem when it is a composite?

Reply:
> not a fan of this merge method. It extracts relatively low level java.time.DateTimeFormatter when it should stick with org.elasticsearch.common.time.DateFormatter abstraction as long as possible.

Maybe we can refactor this in a separate PR then? Also, this code should only be there temporary? As soon as we get rid of Joda time in the code base, I expect that we can get rid of quite a few abstractions.

> I suspect that roundUpBuilder will suffer from the same problem when it is a composite?

Is it intended to be used as a composite?

Reply:
> Is it intended to be used as a composite?

 It is used as a composite in the `merge` method when constructing the roundUpParser with the `appendOptional`. 
I can imagine we can have the same pattern as on the issue (`yyyy-MM-dd'T'HH:mm:ss||yyyy-MM-dd'T'HH:mm:ss.SSS`) but used with this parser. Will try to come up with a testcase

Reply:
I discussed that with @spinscale and there is no way we could create a pattern that would suffer from the same problem.
For index name calculations like `prefix-{2010-01-01/d{yyyy-MM-dd||yyyy-MM-ddTHH}}` it would fail parsing, as the only thing expected after `|` is the timezone (it would fail saying unexpected `|`)